### PR TITLE
Lazily evaluate the IAM Role

### PR DESCRIPTION
**Pull request by robmorgan**
## Description

Lazily evaluate the IAM Role. At the moment we are evaluating Terragrunt configurations for the IAM Role parameters regardless of whether the CLI args are set. When running Terragrunt commands against a large repo, the performance penalty quickly adds up.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

**Comment by robmorgan** [comment]
Hey @denis256, I wonder if I can improve how we evaluate the IAM roles with this PR? At the moment we partially evaluate a Terragrunt configuration for the IAM role regardless of whether or not the CLI args are set. This happens for every file. I'm aware of https://github.com/gruntwork-io/terragrunt/issues/1853, but am I actually regressing anything with this PR?

The one catch with this approach is that any missing IAM CLI arguments won't get merged with what's in a Terragrunt configuration.

File: config/config.go
Comment (robmorgan) [inline config/config.go:765]: I might need to remove this, as OriginalIAMRoleOptions will likely be empty if we fall into this conditional block.

Comment (denis256) [reply config/config.go:765]: AFAIK, OriginalIAMRoleOptions can be populated from CLI ARGs

**Comment by denis256** [comment]
So far I thought about 2 approaches that require more refinement/tests:
* parallel evaluation of HCL files 
* lazy evaluation of IAM role (and other fields too) - single parsing of HCL files and fetching of "promises" for fields which are evaluated(once and cached) on reading value

File: config/config.go
Comment (robmorgan) [reply config/config.go:765]: So I guess the other fields will be overridden unless `RoleARN` is set from the CLI args.

**Comment by robmorgan** [comment]
@denis256 

> So far I thought about 2 approaches that require more refinement/tests:
>parallel evaluation of HCL files
> lazy evaluation of IAM role (and other fields too) - single parsing of HCL files and fetching of "promises" for fields which are evaluated(once and cached) on reading value

Ok nice! Do you think they'll take a while to implement? I guess I'm primarily trying to see if we can find a quick win that is mostly backwards compatible. At the moment on some basic benchmarks, we're spending about ~15% of the CPU budget in the `config.setIAMRole` function. I was hoping to dramatically reduce that if the user specifies one (`--terragrunt-iam-role`) or all of the IAM CLI args.

Screenshot 2023-04-03 at 7 35 58 pm

**Comment by denis256** [comment]
It is not easy a simple change - both approaches will require a significant change in the handling of HCL files

**Comment by robmorgan** [comment]
@denis256 okay just so I gather this correctly. We cannot merge this PR "as-is", even if the tests pass? Do I need to take a different approach?

**Comment by robmorgan** [comment]
@denis256 hey Denis, just a friendly bump on the message above.

**Comment by denis256** [comment]
I think with implemented changes, will be good enough to validate through existing integration tests, which seems to pass now

**Review by denis256 — APPROVED** [review]

**Comment by robmorgan** [comment]
@denis256 okay, thats good to hear! Hopefully I haven't regressed anything.

**Pull request merged**
